include/rocksdb/sst_partitioner.h

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+// Copyright (c) 2012 The LevelDB Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file. See the AUTHORS file for names of contributors.
+//
+// Class for specifying user-defined functions which perform a
+// transformation on a slice.  It is not required that every slice
+// belong to the domain and/or range of a function.  Subclasses should
+// define InDomain and InRange to determine which slices are in either
+// of these sets respectively.
+
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "rocksdb/rocksdb_namespace.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+class Slice;
+
+/*
+ * A SstPartitioner is a generic pluggable way of defining the partition
+ * of SST files. Compaction job will split the SST files on partition boundary
+ * to lower the write amplification during SST file promote to higher level.
+ */
+class SstPartitioner {
+ public:
+  virtual ~SstPartitioner(){};
+
+  // Return the name of this partitioner.
+  virtual const char* Name() const = 0;
+
+  // Called with key that is right after the key that was stored into the SST
+  // Returns true of partition boundary was detected and compaction should
+  // create new file.
+  virtual bool ShouldPartition(const Slice& key) = 0;
```

Comment:
shall this method also take current output file size as parameter, so the partitioner can also split SST by size?

Reply:
Split by size is functionality that compaction knows and there is a lot of options around it. Do we really want to introduce another way? Isn't it confusing? Main purpose of this change is to split based on data.

Reply:
For our use case splitting SST is just a heuristic, not a mandatory. I want to be more flexible in that I can do "split SST based on key if the file is not too small or too large, otherwise override the behavior by size".

Reply:
It mean to say "please partition now".

Reply:
It is "please partition now". I have put there two parameters to make it easier (removed reset)